lib/mail/delivery_recorder.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+module Mail
+  class DeliveryRecorder
+    def self.setup!(enabled:)
+      Mail::Message.include(MessageExtension)
+      ActionMailer::Base.register_observer(new(enabled: enabled))
+    end
+
+    def initialize(enabled:)
+      @enabled = enabled
+    end
+
+    def delivered_email(mail)
+      return unless enabled?
+
+      response = mail.delivery_method.response
+
+      ApplicationRecord.transaction do
+        email = Email.create!(
+          id: response.id,
+          from: response.content["from_email"],
+          to: mail.original_to,
+          template_id: response.template["id"],
+          template_version: response.template["version"],
+          uri: response.uri,
+          personalisation: mail.header["personalisation"].unparsed_value,
+        )
+
+        email.associate_with(*User.where(email: email.to).to_a)
```

Comment:
Should we add `as: "recipient"` or something like that here?

Reply:
Hmmm, the idea behind `as` in association was to give the name to the email itself, not to the associated object role for that email... But having both would be actually nice, especially as the email can be, theoretically, associated with two object of the same class but in different role (for example a hypothetical email from one User regarding another user).

How would you see the following change in the interface here:
```
template_mail(...).tag(:request_for_participant_details).associate_with(user, as: :recipient).associate_with(profile, as: :participant_profile)
```

Justification - `tag` would be used to describe the content of the email and would be stored on the Email record. For example, we have two `request_for_participant_details` email - the initial invitation to the system. We might want to see all the request we sent, but the initial email might have additional `invitation` tag (so `tags` would be an array). Association itself would only store the role given object plays for the object (so 'recipient` `school_cohort` etc) - it is non-nullable and (probably) defaults to an underscored class name of the object. When email is displayed in admin we could then display all the links to associated objects.